src/dataflow/src/render/sources.rs

```diff
@@ -88,10 +88,39 @@ where
             // Create a new local input (exposed as TABLEs to users). Data is inserted
             // via SequencedCommand::Insert commands.
             SourceConnector::Local(_) => {
-                let ((handle, capability), stream) = scope.new_unordered_input();
-                render_state
-                    .local_inputs
-                    .insert(src_id, LocalInput { handle, capability });
+                // WIP is src_id or orig_id the right thing to pass in here?
+                let persisted_source = if let Some(persist) = &mut render_state.persist {
+                    if let Some(stream_name) = persist.stream_name(src_id, &src) {
+                        let token = persist
+                            .persister
+                            .create_or_load(&stream_name)
+                            .expect("WIP this could be an error if the persistence runtime has unexpectedly shut down, what do we do with this?");
+                        let ((handle, capability), stream) =
+                            scope.new_persistent_unordered_input(token);
+                        let stream = stream.map(|(row, ts, diff)| {
+                            let row = persistcfg::string_to_row(&row)
+                                .expect("WIP this should be emitted in err_collection");
```

Comment:
got nerdsniped into making this work I think this is close to right (only tested that it compiled)

```
diff --git a/src/dataflow-types/src/errors.rs b/src/dataflow-types/src/errors.rs
index f0b137e72..60e8e0f5c 100644
--- a/src/dataflow-types/src/errors.rs
+++ b/src/dataflow-types/src/errors.rs
@@ -49,6 +49,7 @@ impl Display for SourceError {
 pub enum SourceErrorDetails {
     Initialization(String),
     FileIO(String),
+    Persistence(String),
 }

 impl Display for SourceErrorDetails {
@@ -62,6 +63,7 @@ impl Display for SourceErrorDetails {
                 )
             }
             SourceErrorDetails::FileIO(e) => write!(f, "file IO: {}", e),
+            SourceErrorDetails::Persistence(e) => write!(f, "persistence: {}", e),
         }
     }
 }
diff --git a/src/dataflow/src/render/sources.rs b/src/dataflow/src/render/sources.rs
index 340b08efd..7e053b136 100644
--- a/src/dataflow/src/render/sources.rs
+++ b/src/dataflow/src/render/sources.rs
@@ -97,22 +97,40 @@ where
                             .expect("WIP this could be an error if the persistence runtime has unexpectedly shut down, what do we do with this?");
                         let ((handle, capability), stream) =
                             scope.new_persistent_unordered_input(token);
-                        let stream = stream.map(|(row, ts, diff)| {
-                            let row = persistcfg::string_to_row(&row)
-                                .expect("WIP this should be emitted in err_collection");
-                            (row, ts, diff)
-                        });
-                        Some(((LocalInputHandle::Persistent(handle), capability), stream))
+                        let (ok_stream, err_stream) =
+                            stream.map_fallible(move |(row, ts, diff)| {
+                                persistcfg::string_to_row(&row)
+                                    .map(|row| (row, ts, diff))
+                                    .map_err(|e| {
+                                        SourceError::new(
+                                            stream_name.clone(),
+                                            SourceErrorDetails::Persistence(e),
+                                        )
+                                    })
+                            });
+                        Some((
+                            (LocalInputHandle::Persistent(handle), capability),
+                            ok_stream,
+                            err_stream
+                                .map(DataflowError::SourceError)
+                                .pass_through("table-errors")
+                                .as_collection(),
+                        ))
                     } else {
                         None
                     }
                 } else {
                     None
                 };
-                let ((handle, capability), stream) = persisted_source.unwrap_or_else(|| {
-                    let ((handle, capability), stream) = scope.new_unordered_input();
-                    ((LocalInputHandle::Transient(handle), capability), stream)
-                });
+                let ((handle, capability), ok_stream, err_collection) = persisted_source
+                    .unwrap_or_else(|| {
+                        let ((handle, capability), stream) = scope.new_unordered_input();
+                        (
+                            (LocalInputHandle::Transient(handle), capability),
+                            stream,
+                            Collection::empty(scope),
+                        )
+                    });

                 render_state.local_inputs.insert(
                     src_id,
@@ -122,12 +140,11 @@ where
                     },
                 );
                 let as_of_frontier = self.as_of_frontier.clone();
-                let ok_collection = stream
+                let ok_collection = ok_stream
                     .map_in_place(move |(_, mut time, _)| {
                         time.advance_by(as_of_frontier.borrow());
                     })
                     .as_collection();
-                let err_collection = Collection::empty(scope);
                 self.insert_id(
                     Id::Global(src_id),
                     crate::render::CollectionBundle::from_collections(
```